### ISSUE TITLE
DL models_ЧП6

### ISSUE THREAD

**Issue by DariGD** [DariGD/ai24-toxicity-prediction#38]
## Шаги

- [ ] Разработать NN для предикта метрики токcичности LD50
- [ ] Сравнить результаты DL-модели с предыдущими полученными результатами ML-моделей
- [ ] Результаты зафиксировать в отдельном md-файле

**Comment by DariGD** [comment]
- Разработаны различные архитектуры нейронной сети
- Определена наилучшая - глубокая полносвязная регрессионная модель на PyTorch с 5 линейными слоями (функция активации ReLU, Dropout после первых трёх скрытых слоёв с вероятностью 0.2 для регуляризации, выходная активация — Softplus)
- Проведен эксперимент по анализу изменений качества применяемой модели при изменении диапазона таргета. Результат - удалось добиться улучшения MAPE при введении ограничений по таргету не менее 10 мг/кг.

| Модель | MAPE (прямое значение из sklearn до умножения на 100!) |
|------------------|------------------|
| LinearRegression            | 2.94  на кросс-валидации   | 
| XGBoost             | 2.64 на кросс-валидации       | 
| NN (5 слоев)             | 6.71 на тесте       | 
| NN (5 слоев) на таргете >10            | 0.88 на тесте     | 


**Основные выводы**

- Наилучший MAPE на тестовой выборке (на всем объеме данных) - 6.71. По эпохам - в ноутбуке
- Наилучший MAPE на тестовой выборке (на таргете > 10) - 0.88. По эпохам - в ноутбуке
- MAPE крайне чувствительная метрика к маленьким значениями, поэтому на диапазонах от 0 до 10-20 отклонения предикта вносят сильный вклад в MAPE и он "взрывается"
- Возможно имеет смысл изменить метрику качества для данных с тагретом до 10-50 (примерный диапазон) и обучать отдельную модель на них, чтобы максимально точно предсказывать, как высокие значения (с чем текущие модели справляются), так и небольшие до 10 мг/кг с одинаково высокой точностью. Добавить промежуточную модель для классификации.


Ноутбук - https://github.com/DariGD/ai24-toxicity-prediction/blob/16ba8ef9dda47c62d5465beaed2204482cada9ce/notebooks/NN_models_project.ipynb

**Comment by bugor40** [comment]
Очень круто! Молодец!

**Comment by Kikongo** [comment]
Были обучены различные модели нейронных сетей.

* В ноутбуке содержится несколько вариантов нейронных сетей с различным количеством слоев, нейронов, функциями активации (ReLU, Sigmoid), а также с применением Dropout и Batch Normalization.
* Экспериментирование с разными оптимизаторами и скоростями обучения.
* Исследование влияния отбора признаков на качество модели.

| Модель | MAPE |
|------------------|------------------|
| LinearRegression            | 41.4   | 
| XGBoost             | 2.29     | 
| СatBoost           | 4.31     | 
| NN      | 4.09     | 

В целом, нейронные модели показывают неплохие результаты. Есть смысл дальше с ними работать

Ноутбук: https://github.com/DariGD/ai24-toxicity-prediction/blob/Models/notebooks/DL_tox.ipynb

**Comment by VladislavTokarev02** [comment]
В ходе работы не удалось добиться желаемого результата, наилучший MAPE = 9.95. 

Наиболее удачная архитектура: 
- 5 слоёв;
- DropOut, AlphaDropOut, weight_decay;
- GeLU;
- BatchNorm.

Ссылка: https://github.com/DariGD/ai24-toxicity-prediction/blob/Models/notebooks/no_metadata_tvp.ipynb
